src/test/java/com/google/sps/ChatServletTest.java

```diff
@@ -0,0 +1,258 @@
+package com.google.sps;
+
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.verify;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.sps.servlets.ChatServlet;
+import com.google.template.soy.SoyFileSet;
+import com.google.template.soy.tofu.SoyTofu;
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.Scanner;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+@RunWith(JUnit4.class)
+public final class ChatServletTest extends Mockito {
+    private static final String GROUP_ID_PARAM = "groupID";
+    private static final String GROUP_ID = "123";
+    private static final String MESSAGE_TEXT_TOXIC = "what kind of idiot name is foo?";
+    private static final String MESSAGE_KIND = "Message-";
+    private static final String MESSAGE_TEXT_PROPERTY = "message-text";
+    private static final String TIMESTAMP_PROPERTY = "timestamp";
+    private static final String GROUP_KIND = "Group";
+    private static final String GROUP_NAME_PROPERTY = "name";
+    private static final String MESSAGES_MAP_KEY = "messages";
+    private static final String GROUPS_KEY = "groups";
+    private static final ImmutableList<String> SAMPLE_GROUP_LIST = 
+        ImmutableList.of("Black Lives Matter");
+    private static final String CHAT_TEMPLATE = "templates.chat.chatPage";
+    private static final Double SCORE_OFFSET = 0.0000005;
+    private static final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+    private File apiKeyFile;
+    private Scanner scanner;
+    private String apiKey;
+    private String referer;
+    private String apiURL;
+    private StringWriter stringWriter;
+    private PrintWriter printWriter;
+    private SoyFileSet sfs;
+    private SoyTofu tofu;
+    private DatastoreService datastore;
+    private ImmutableMap<String, ImmutableList<String>> templateData;
```

Comment:
If a variable is only used in one test, it can be local. In addition, if a variable doesn't need to share state across tests, or isn't initialized in setup then it should be a local variable.